### PR TITLE
New class "GServer"

### PR DESCRIPTION
**Pull request by annando**
The GServer class now contains functions to analyze a given server address and to check the used software, version and other data.

This is a replacement for several functions in the "PortableContact" class.

This is the first step in the afford to reduce and rework that class file.

### CODE REVIEW

File: src/Model/GServer.php [hunk 1]
```diff
@@ -0,0 +1,1148 @@
...
+
+		$serverdata = self::checkPoCo($url, $serverdata);
+
+		// We can't detect the network type. Possibly it is some system that we don't know yet
+		if (empty($serverdata['network'])) {
+			$serverdata['network'] = Protocol::PHANTOM;
+		}
+
+		// When we hadn't been able to detect the network type, we use the hint from the parameter
+		if (($serverdata['network'] == Protocol::PHANTOM) && !empty($network)) {
```
Comment (MrPetovan) [inline src/Model/GServer.php]: Inside parentheses are unnecessary.

File: src/Model/GServer.php [hunk 2]
```diff
@@ -0,0 +1,1148 @@
...
+			$serverdata['network'] = $network;
+		}
+
+		// Check host-meta for phantom networks.
+		// Although this is not needed, it is a good indicator for a living system,
+		// since most systems had implemented it.
+		if (($serverdata['network'] == Protocol::PHANTOM) && !self::validHostMeta($url)) {
```
Comment (MrPetovan) [inline src/Model/GServer.php]: Inside parentheses are unnecessary.

File: src/Model/GServer.php [hunk 3]
```diff
@@ -0,0 +1,1148 @@
...
+		}
+
+		$serverdata['url'] = $url;
+		$serverdata['nurl'] = Strings::normaliseLink($url);
+
+		// We take the highest number that we do find
+		$registeredUsers = $serverdata['registered-users'] ?? 0;
+
+		// On an active server there has to be at least a single user
+		if (($serverdata['network'] != Protocol::PHANTOM) && ($registeredUsers == 0)) {
```
Comment (MrPetovan) [inline src/Model/GServer.php]: Inside parentheses are unnecessary.

File: src/Model/GServer.php [hunk 4]
```diff
@@ -0,0 +1,1148 @@
...
+		if (!is_array($data)) {
+			return;
+		}
+
+		$gserver = DBA::selectFirst('gserver', ['id', 'relay-subscribe', 'relay-scope'], ['nurl' => Strings::normaliseLink($server_url)]);
+		if (!DBA::isResult($gserver)) {
+			return;
+		}
+
+		if (($gserver['relay-subscribe'] != $data['subscribe']) || ($gserver['relay-scope'] != $data['scope'])) {
```
Comment (MrPetovan) [inline src/Model/GServer.php]: Inside parentheses are unnecessary.

File: src/Model/GServer.php [hunk 5]
```diff
@@ -0,0 +1,1148 @@
...
+	}
+
+	/**
+	 * Parses Nodeinfo 2
+	 *
+	 * @param string $nodeinfo_url address of the nodeinfo path
+	 * @return array Server data
+	 * @throws \Friendica\Network\HTTPException\InternalServerErrorException
+	 */
+	private static function parseNodeinfo2(string $nodeinfo_url)
```
Comment (MrPetovan) [inline src/Model/GServer.php]: There is a bunch of duplicated code between this method and `parseNodeinfo1`. Ideally, we should have a `Protocol\NodeInfoV2` class extending `Protocol\NodeInfoV1` that would include the common extraction methods to both versions.

File: src/Model/GServer.php [hunk 6]
```diff
@@ -0,0 +1,1148 @@
...
+			}
+		}
+
+		$serverdata['registered-users'] = max($serverdata['registered-users'], count($contacts));
+
+		return $serverdata;
+	}
+
+	/**
+	 * Checks if the given server does have a "poco" endpoint
```
Comment (MrPetovan) [inline src/Model/GServer.php]: Please explicit the acronym PoCo for Portable Contact in the comment.

File: src/Model/GServer.php (hunk 1 above)
Comment (annando) [reply src/Model/GServer.php]: But they increase readability.

File: src/Model/GServer.php (hunk 5 above)
Comment (annando) [reply src/Model/GServer.php]: I hadn't found a good way that wouldn't reduce the readability.

File: src/Model/GServer.php (hunk 1 above)
Comment (MrPetovan) [reply src/Model/GServer.php]: We don't have the same personal definition of "readability". 😅

Comment (annando) [reply src/Model/GServer.php]: Obviously 😀

**Pull request merged**